### PR TITLE
[Merged by Bors] - feat: port Algebra.Category.ModuleCat.Projective

### PR DESCRIPTION
**Pull request by apurvnakade**
This is stuck at a universe issue. Lean is having trouble merging `Type (max (max v u) u v)` with `Type (max u u v)`

See comment below. 

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Algebra/Category/Module/Projective.lean [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+
+/-- The categorical notion of projective object agrees with the explicit module-theoretic notion. -/
+theorem IsProjective.iff_projective {R : Type u} [Ring R] {P : Type max u v} [AddCommGroup P]
+    [Module R P] : Module.Projective R P ↔ Projective (ModuleCat.of R P) := by
+  refine' ⟨fun h => _, fun h => _⟩
+  · letI : Module.Projective R (ModuleCat.of R P) := h
+    exact ⟨fun E X epi => Module.projective_lifting_property _ _ ((ModuleCat.epi_iff_surjective _).mp epi)⟩
+  · refine' Module.Projective.of_lifting_property.{u,v} _
+    intro E X mE mX sE sX f g s
+    haveI : Epi (ModuleCat.asHom f) := (ModuleCat.epi_iff_surjective (ModuleCat.asHom f)).mpr s
```
Comment (apurvnakade) [inline Mathlib/Algebra/Category/Module/Projective.lean:39]: For some reason the operator `↟` is not working.

Comment (semorrison) [reply Mathlib/Algebra/Category/Module/Projective.lean:39]: I can't find `↟` in the source? If you delete stuff from the mathport output, please put in comments showing the original code!

Comment (semorrison) [reply Mathlib/Algebra/Category/Module/Projective.lean:39]: `open ModuleCat`

File: Mathlib/Algebra/Category/Module/Projective.lean [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+    letI : Projective (ModuleCat.of R P) := h
+    exact ⟨Projective.factorThru (ModuleCat.asHom g) (ModuleCat.asHom f),
+      Projective.factorThru_comp (ModuleCat.asHom g) (ModuleCat.asHom f)⟩
+#align is_projective.iff_projective IsProjective.iff_projective
+
+namespace ModuleCat
+
+variable {R : Type u} [Ring R] {M : ModuleCat.{max u v} R}
```
[the remark this answers is not in the document]
Comment (semorrison) [reply Mathlib/Algebra/Category/Module/Projective.lean:47]: Search for `TopCatMax`, see if you can imitate that, and ask again if it is not making sense! :-)

File: Mathlib/Algebra/Category/Module/Projective.lean [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+set_option linter.uppercaseLean3 false in
+#align Module.projective_of_free ModuleCat.projective_of_free
+
+/-- The category of modules has enough projectives, since every module is a quotient of a free
+    module. -/
+
+instance moduleCat_enoughProjectives : EnoughProjectives (ModuleCat.{max u v} R)
+    where presentation M :=
+    ⟨{  p := ModuleCat.of R (Finsupp R M) --(M →₀ R)
+        projective := projective_of_free.{u, v} Finsupp.basisSingleOne.{u, v}
```
Comment (apurvnakade) [inline Mathlib/Algebra/Category/Module/Projective.lean:68]: Another universe error :-|

Comment (apurvnakade) [reply Mathlib/Algebra/Category/Module/Projective.lean:68]:
```lean 

application type mismatch
  projective_of_free Finsupp.basisSingleOne
argument
  Finsupp.basisSingleOne
has type
  Basis ?m.10039 ?m.10038 (?m.10039 →₀ ?m.10038) : Type (max (max v u) u v)
but is expected to have type
  Basis (?m.10185 M) R ↑(of R (R →₀ ↑M)) : Type (max u u v)
```

Comment (apurvnakade) [reply Mathlib/Algebra/Category/Module/Projective.lean]: Thank you!

**Comment by ChrisHughes24** [comment]
bors merge

**Pull request closed without merging**